### ISSUE TITLE
Set up automatic execution of training material WFs against main using Jenkins and Planemo. This [tutorial](http://galaxyproject.github.io/training-material/topics/transcriptomics/tutorials/srna/tutorial.html) is a good start.

### ISSUE THREAD

**Issue by nekrut** [galaxyproject/usegalaxy-playbook#164]

**Comment by erasche** [comment]
We have some stuff for this already! https://github.com/usegalaxy-eu/workflow-testing

We're going to expand it to include automatic testing of all workflows from the training materials, but had not had the time to do that due to other more time sensitive issues.

**Comment by martenson** [comment]
@erasche yeah, we were hoping to piggyback :)

**Comment by erasche** [comment]
@martenson awesomesauce!

**Comment by luke-c-sargent** [comment]
current status:
 - [forked](https://github.com/luke-c-sargent/workflow-testing) the linked repo from @erasche (thanks!)
 - added a Jenkinsfile and registered a Jenkins job named `automated workflow testing` in the `test_jobs` tab of the job list
 - adapted the .travis.yml's approach of getting a [planemo test script](https://raw.githubusercontent.com/galaxyproject/planemo/master/scripts/run_galaxy_workflow_tests.sh), setting the env, and providing a workflow file
 - running against test.galaxyproject.org

some issues:
 - Jenkins job is failing as it is unable to see the credentials i created for that purpose; there are probably some scoping issues i am not getting.
 - not really sure whats happening with the actual test; as best i can tell, passing the workflow file to the script anchors it in the workflow's directory, making planemo aware of the .yml job and test files (which i have yet to write for my trivial example)

any clarification, suggestions or well-meaning admonishments welcome

**Comment by bgruening** [comment]
Hi @luke-c-sargent!

How does your jenkins script look like? We are running this here:

```
pip install https://github.com/galaxyproject/planemo/archive/master.zip
pip install galaxy-parsec
export PLANEMO_OPTIONS="--verbose"
bash fetch-urls.sh
bash run_galaxy_workflow_tests.sh $WORKFLOW
```

`WORKFLOW` is the path to the folder of every workflow in this case.

**Comment by erasche** [comment]
If you use our `run_galaxy_workflow_tests.sh` script there are a couple of fun bonuses:

- We write out some HTML files that [are available to users](https://build.galaxyproject.eu/job/usegalaxy.eu/job/workflow-testing/PYTHON=System-CPython-2.7,WORKFLOW=example1%2Fwf3-shed-tools.ga/)
  - the tool test output is made visible (but this is currently quite bare in planemo)
  - And we generate an IFrame with a link to the history that was created, so you can see the actual tools that were executed, which is sometimes useful.

**Comment by luke-c-sargent** [comment]
using `run_galaxy_workflow_tests.sh` on example1, i get `"execution_problem": "'file_name'"` -- not an especially telling error.

cmd: `./run_galaxy_workflow_tests.sh example1/wf3-shed-tools.ga`

this creates a history, _e.g._ `example1/wf3-shed-tools.ga 1540243522`, and uploads the precursor datasets `Genes.bed` & `GSE37268_mof3.out.hpeak.txt`, then throws the above error without executing the workflow. i feel like i must be missing something obvious.

thanks!

(running on Mac OS)

**Comment by erasche** [comment]
Have you tried running just the planemo command?

```bash
planemo test \
	--galaxy_url "https://usegalaxy.eu" \
	--galaxy_user_key "$GALAXY_USER_KEY" \
    --no_shed_install \
	--engine external_galaxy \
    example1/wf3-shed-tools.ga
```

**Comment by luke-c-sargent** [comment]
same deal :(
`Test case [<planemo.runnable.TestCase object at 0x1053009d0>] resulted in run response [Run failed with message ['file_name']]`

from
```
planemo -v test \
        --history_name "a_test_history" \
        --galaxy_url "https://test.galaxyproject.org" \
        --galaxy_user_key "<REDACTED>" --no_shed_install \
        --engine external_galaxy \
        example1/wf3-shed-tools.ga;
```
which does indeed create a history on `test` with the input data and uploads the requisite workflow. this also happens when testing against main.

also tested `example1` and a custom workflow against usegalaxy.eu (as user `sargentl` if you happen to feel like looking in to it) -- first execution stalled, so i `ctrl+c`'d it, added a `-v` flag and tried again and it went through to completion (the 'stall' could have just been impatience on my part). further executions worked as expected.

 i wonder if it's a 18.09 issue -- what version does usegalaxy.eu run?

**tl;dr: usegalaxy.org & test.galaxyproject.org give this `file_name` error, usegalaxy.eu works with example workflows and my own custom workflow**

**Comment by erasche** [comment]
We run 18.05 still, you can always check at https://usegalaxy.eu/api/version

Sounds like potentially an 18.09 specific issue.

**Comment by luke-c-sargent** [comment]
to delve deeper i ran two local versions of galaxy (18.05 and 18.09) and both threw:
```
Test case [<planemo.runnable.TestCase object at 0x10549c9d0>] resulted in run response [Run failed with message ['file_name']]
```
error, so that seems to rule out it being a mere 18.09 issue. perhaps it's some sort of race condition? `ping` takes 60ms on usegalaxy.org and 170ms on usegalaxy.eu -- but that seems like a stretch. i will do some local server log digging to see what's what and report back.

**Comment by martenson** [comment]
I am able reproduce the error @luke-c-sargent reported and will debug it next week hopefully with help of @jmchilton since this smells planemo-related.

**Comment by jmchilton** [comment]
There is no chance you are using an API key with EU and not in these other cases?

**Comment by jmchilton** [comment]
Create a gist from the verbose logging Planemo creates?

**Comment by erasche** [comment]
Could not tell if you were addressing it to me/not, but, no, EU's testing uses a non-privileged user.

**Comment by martenson** [comment]
@jmchilton I will get the logging to you, give me a few

**Comment by martenson** [comment]
@jmchilton https://jenkins.galaxyproject.org/job/workflow_testing/PYTHON=System-CPython-2.7,TESTED_WF_PATH=example1%2Fwf3-shed-tools.ga/10/consoleFull

**Comment by jmchilton** [comment]
It is dying on this in the Planemo client I believe...

```python
    galaxy_paths = []
    for (dataset, upload_target) in datasets:
        if isinstance(upload_target, FileUploadTarget):
            local_path = upload_target.path
            ctx.vlog("fetching full dataset for %s, %s" % (dataset, local_path))
            dataset_full = user_gi.datasets.show_dataset(dataset["id"])
            galaxy_path = dataset_full["file_name"]
            ctx.vlog("galaxy_path is %s" % galaxy_path)
            job_path = os.path.join(job_dir, local_path)
            galaxy_paths.append((job_path, galaxy_path))
```

**Comment by jmchilton** [comment]
... which hasn't been used since https://github.com/galaxyproject/planemo/commit/9ab4a0d54a748fd47f560094aaff79917129c7f9#diff-dd8266ef5b6bb7cbed33e52f9a0ab5cc.

**Comment by jmchilton** [comment]
https://github.com/galaxyproject/planemo/pull/882

**Comment by jmchilton** [comment]
Released a new Planemo with that fix. Next bug 😉?

**Comment by martenson** [comment]
It is executing all the tools now, yay!

https://jenkins.galaxyproject.org/job/workflow_testing/PYTHON=System-CPython-2.7,TESTED_WF_PATH=example1%2Fwf3-shed-tools.ga/14/console

**Comment by martenson** [comment]
The job is now executing 4 wf tests against Test and publishes HTML report on each. One test is passing, others failing for various reasons, but I suspect none of the reasons is our deployment's problem but rather the test cases do not seem to be very robust.

https://jenkins.galaxyproject.org/job/workflow_testing/

Nevertheless as our next step @natefoo is going to get us a config-copy of Main running on independent database (so we do not pollute) and we will set these jobs to run there. 

Until then we should gather a reliable and useful set of workflows to test (and test them).

**Comment by bgruening** [comment]
> Released a new Planemo with that fix. Next bug wink?

@jmchilton not a bug, but we have a workflow with a composite datatype as input. Would this be hard to implement? Not super urgent but would make some of our proteomic people happy. Thanks John!

**Comment by luke-c-sargent** [comment]
current status of this on `main`:
@erasche's [galaxy-eu workflow testing repo](https://github.com/usegalaxy-eu/workflow-testing)  we wish to leverage continues to expand with additional training workflow invocations; i have gotten these tests going on Jenkins against both main and usegalaxy.eu to varying levels of success. in running these against main, an issue regarding tool version has appeared:

running `wf3-shed-tools.ga` from example 1 requires version 1.1.0 of tool `tp_replace_in_column`, which is not on main. The error reports the absense of the upgraded version:
```
{"err_msg": "Tool toolshed.g2.bx.psu.edu/repos/bgruening/text_processing/tp_replace_in_column/1.1.2 missing. Cannot add dummy datasets.", "err_code": 400014}
```
...[which main has](https://usegalaxy.org/?tool_id=toolshed.g2.bx.psu.edu%2Frepos%2Fbgruening%2Ftext_processing%2Ftp_replace_in_column%2F1.1.2&version=1.1.2&__identifer=cfj1ijsx8bm).

So, either the workflow invocation logic is upgrading the tool without success (bad for reproducibility -- different tool version -> porentially different results)  or is reporting absence of the tool but mistakenly reports the most recent version on the server.

Either way, this begs the question: in the case of a training workflow with an outdated/missing tool version: should the training workflow be updated, or should an older tool version be added to the server?

Looking forward, most of the existent workflows [run and succeed via Jenkins](https://jenkins.galaxyproject.org/view/test_jobs/job/training-workflow-test/) against `usegalaxy-eu` with a couple exceptions -- the next step is investigate why the failures failed. once fixed, i will run these tests against main to note any disparities like the versioning issue above; from there, i can help with the real heart of the matter, actually expanding upon erasche's list of workflow files and their provisioning yml configs.

**Comment by martenson** [comment]
At the moment I think we are in a good shape regarding the jenkins job and testing contents. We discussed with @natefoo and think that now we need a non-Main instance against which we can test. We decided we will not use Main to prevent non-critical database increase -especially given that we can mirror all the configs, just spare the database from the test data.

@luke-c-sargent you can stand by, I will ping you here when we have the instance up, thanks a lot!
